### PR TITLE
Support for distributed outputs as driver responses; support for group approximation across distributed components in certain cases.

### LINKED ISSUE #1331: Cannot use outputs of distributed components as pyoptsparse objectives or constraints
### Summary of Issue

While developing on a distributed component in another project, I found a couple of undocumented (I think) things about the distributed component implementation.

1) Setting the output of a distributed component as either a constraint or objective of an optimization problem does not work correctly (at least with pyoptsparsedriver) when running in parallel

2) Check_partials and check_totals do not appear to work correctly when running parallel

### Issue Type

- [ X] Bug
- [ X] Enhancement  

### Description

Adding an output of a distributed component as an optimization constraint does not work (produces a pyoptsparse error message) although list_outputs and other openmdao-based interactions display correctly concatenated distributed arrays.

See the gist below.

Also, check_totals and check_partials appear to be broken right now when run in parallel on a distributed component which makes developing parallel components much trickier. If only non-distributed comps are include= in the check_partials, then it works, so the problem is not about parallel in general, but about the distributed outputs.

### Example

See the gist here: [link](https://gist.github.com/bbrelje/b6ac496b96c1ce1d393226078314d25d)

When run under mpi with n=2 or more, the GroupThatDoesNotWork() the following error is raised:
```
`+------------------------------------------------------------------------------+
| pyOptSparse Error: 1 constraint values were returned in distcomp.y, but      |
| expected 2.                                                                  |
+------------------------------------------------------------------------------+

Traceback (most recent call last):
  File "toy.py", line 88, in <module>
    prob.run_driver()
  File "/usr/local/lib/python2.7/dist-packages/openmdao/core/problem.py", line 592, in run_driver
    return self.driver.run()
  File "/usr/local/lib/python2.7/dist-packages/openmdao/drivers/pyoptsparse_driver.py", line 410, in run
    storeHistory=self.hist_file, hotStart=self.hotstart_file)
  File "/usr/local/lib/python2.7/dist-packages/pyoptsparse/pySLSQP/pySLSQP.py", line 281, in __call__
    self._waitLoop()
  File "/usr/local/lib/python2.7/dist-packages/pyoptsparse/pyOpt_optimizer.py", line 634, in _waitLoop
    self._masterFunc2(*info)
  File "/usr/local/lib/python2.7/dist-packages/pyoptsparse/pyOpt_optimizer.py", line 352, in _masterFunc2
    fcon = self.optProb.processConstraints(funcs)
  File "/usr/local/lib/python2.7/dist-packages/pyoptsparse/pyOpt_optimization.py", line 1196, in processConstraints
    self.constraints[iCon].ncon))
pyoptsparse.pyOpt_error.Error`
```

When run under mpi with n=1, the optimisation succeeds. 

If check_partials is uncommented and run in parallel, the following error is raised:

```
    approximation.compute_approximations(comp, jac=approx_jac)
  File "/usr/local/lib/python2.7/dist-packages/openmdao/approximation_schemes/finite_difference.py", line 200, in compute_approximations
    self._compute_approximations(system, jac, total, system._outputs._under_complex_step)
  File "/usr/local/lib/python2.7/dist-packages/openmdao/approximation_schemes/approximation_scheme.py", line 424, in _compute_approximations
    oview[:, i] = result
ValueError: could not broadcast input array from shape (2) into shape (1)
```

### Environment
Operating System: Ubuntu 16.04
Python environment: system Py2 and Anaconda Py3
Packages: OM 2.10.1 with Py2, 
OM 3.0.0/Numpy 1.16.4/pyoptsparse 2.0.3 on Python 3

### LINKED ISSUE #1378: Fix group 'fd' and check_totals to work with distributed components.
### Summary of Issue

Fix group 'fd' and check_totals to work with distributed components.

### Issue Type

- [x ] Bug
- [ ] Enhancement  
      POEM #___ <N/A or POEM number at https://github.com/OpenMDAO/POEMs>

### Description

Fix group 'fd' and check_totals to work with distributed components.

### Environment

Operating System: <i.e. OS X 10.14.6, Windows 10, Ubuntu 16.04>

Python environment: <i.e. Anaconda Python 3.7.1>

Packages: <versions of Dymos, OpenMDAO, and other pertinent Python packages. You can just paste the output of `pip freeze` or `conda list` here.>

### PR DESCRIPTION
**Pull request by Kenneth-T-Moore**
### Summary

Adds support for using distributed outputs as driver responses to PyoptsparseDriver and ScipyOptimizer.
Adds support for group FD/CS at the top of a model containing a distributed component.
Adds support for `check_totals` to work on a model containing a distributed component.
Adds support for group FD/CS of a subgroup containing a distributed component provided that there are no distributed inputs entering the group.

### Related Issues

- Resolves #1331 
- Resolves #1378 

### Backwards incompatibilities

None

### New Dependencies

None

### CODE REVIEW

File: openmdao/core/driver.py [hunk 1]
```diff
@@ -468,6 +490,17 @@ def _get_voi_val(self, name, meta, remote_vois, driver_scaling=True, rank=None):
                     val = np.empty(size)
 
                 comm.Bcast(val, root=owner)
+
+        elif distributed:
+            local_val = model._get_val(name, flat=True)
+            if indices is not None:
+                val = val[indices]
```
Comment (naylor-b) [inline openmdao/core/driver.py:497]: This val gets overwritten below by `val = np.zeros(np.sum(sizes))`, so indices are not being handled.

File: openmdao/core/tests/test_distrib_derivs.py [hunk 2]
```diff
@@ -307,8 +309,326 @@ def test_fan_in_grouped(self, nlsolver):
...
+        model.add_subsystem('p', ivc, promotes=['*'])
+        model.add_subsystem("parab", DistParab(arr_size=size, deriv_type='dense'), promotes=['*'])
+        model.add_subsystem('sum', om.ExecComp('f_sum = sum(f_xy)',
+                                               f_sum=np.ones((size, )),
+                                               f_xy=np.ones((size, ))),
+                            promotes=['*'])
+
+        model.add_design_var('x', lower=-50.0, upper=50.0)
+        model.add_design_var('y', lower=-50.0, upper=50.0)
+        model.add_constraint('f_xy', lower=0.0)
```
Comment (bbrelje) [inline openmdao/core/tests/test_distrib_derivs.py:332]: I'm a little bit nervous about only having one distributed constraint tested in all cases. Might be good to have an alternate formulation which tracks more than one dist constraint (catches potential nonzero index edge cases)

File: openmdao/core/driver.py [hunk 3]
```diff
@@ -449,9 +465,15 @@ def _get_voi_val(self, name, meta, remote_vois, driver_scaling=True, rank=None):
         vec = model._outputs._views_flat
         indices = meta['indices']
 
+        if MPI:
+            distributed = MPI.COMM_WORLD.size > 0 and name in distributed_vars
```
[the remark this answers is not in the document]
Comment (naylor-b) [reply openmdao/core/driver.py:469]: Not necessarily.

File: openmdao/core/driver.py (hunk 1 above)
Comment (Kenneth-T-Moore) [reply openmdao/core/driver.py:497]: Should be local_val, but this is an obvious coverage hole, so i need to add a test.

File: openmdao/core/driver.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Kenneth-T-Moore) [reply openmdao/core/driver.py:469]: Actually, we use MPI.COMM_WORLD in some places in driver, but we also use the parent problem's comm in other places.  I think maybe we should switch them all over to the parent problem's comm to account for multi problem configurations.

File: openmdao/docs/features/core_features/adding_desvars_objs_consts/adding_constraints.rst [hunk 4]
```diff
@@ -22,3 +22,17 @@ limits in the constraint definition should be specified using these units.
 .. embed-code::
     openmdao.core.tests.test_driver.TestDriverFeature.test_specify_units
     :layout: code, output
+
+
+Using the output of a distributed component as a constraint
```
[the remark this answers is not in the document]
Comment (Kenneth-T-Moore) [reply openmdao/docs/features/core_features/adding_desvars_objs_consts/adding_constraints.rst:27]: Sure, might be better for it to all be in one tutorial-set.  You can move this over if you like, or replace it with a better problem. This one is kind of contrived.

File: openmdao/test_suite/components/paraboloid_distributed.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+        comm = self.comm
+        rank = comm.rank
+
+        sizes, offsets = evenly_distrib_idxs(comm.size, arr_size)
+        start = offsets[rank]
+        io_size = sizes[rank]
+        self.offset = offsets[rank]
+        end = start + io_size
+
+        self.add_input('x', val=np.ones(io_size),
```
[the remark this answers is not in the document]
Comment (Kenneth-T-Moore) [reply openmdao/test_suite/components/paraboloid_distributed.py:36]: Or your could propose a POEM with the new API.

File: openmdao/core/driver.py [hunk 6]
```diff
@@ -449,9 +465,15 @@ def _get_voi_val(self, name, meta, remote_vois, driver_scaling=True, rank=None):
...
+        else:
+            distributed = False
+
         if name in remote_vois:
             owner, size = remote_vois[name]
             # if var is distributed or only gathering to one rank
+            # TODO - support distributed var under a parallel group.
```
[the remark this answers is not in the document]
Comment (Kenneth-T-Moore) [reply openmdao/core/driver.py:476]: That is the mystery.  My prediction is something would hang somewhere. Hope to address in a subsequent story.

File: openmdao/core/tests/test_distrib_derivs.py (hunk 2 above)
Comment (Kenneth-T-Moore) [reply openmdao/core/tests/test_distrib_derivs.py:332]: That is a good point. I'll put together something with a larger set of constraints.

File: openmdao/core/driver.py (hunk 1 above)
Comment (Kenneth-T-Moore) [reply openmdao/core/driver.py]: Looks like defining indices on a distributed response doesn't work yet. I think it makes sense for that to be a separate story, and might require some api discussiion.

Comment (naylor-b) [reply openmdao/core/driver.py]: I think you'll need to treat indices as global in the same way as we treat src_indices, which means you'll have to check what part of the indices refer to a given proc and only gather those from each proc.

**Comment by Kenneth-T-Moore** [comment]
I think I've addressed all of the review comments and have created follow-on stories for a couple of remaining issues.

File: openmdao/core/tests/test_distrib_derivs.py (hunk 2 above)
Comment (Kenneth-T-Moore) [reply openmdao/core/tests/test_distrib_derivs.py:332]: Got a new test added with 2 distributed constraints and 2 regular ones.

**Pull request merged**